synapse/storage/user_directory.py

```diff
@@ -30,7 +28,162 @@
 logger = logging.getLogger(__name__)
 
 
-class UserDirectoryStore(SQLBaseStore):
+class UserDirectoryStore(BackgroundUpdateStore):
+    def __init__(self, db_conn, hs):
+        super(UserDirectoryStore, self).__init__(db_conn, hs)
+
+        self.server_name = hs.hostname
+
+        self.register_background_update_handler(
+            "populate_user_directory", self._populate_user_directory
+        )
+
+    @defer.inlineCallbacks
+    def _populate_user_directory(self, progress, batch_size):
+
+        TEMP_TABLE = "_temp_populate_user_directory_rooms"
+        state = self.hs.get_state_handler()
+
+        # If we don't have progress filed, delete everything.
+        if not progress:
+            yield self.delete_all_from_user_dir()
+
+        # Get all the rooms that we want to process.
+        def _make_staging_area(txn):
+            sql = (
+                "CREATE TABLE IF NOT EXISTS "
+                + TEMP_TABLE
+                + "(room_id TEXT NOT NULL, events BIGINT NOT NULL)"
+            )
+            txn.execute(sql)
+
+            # Get rooms we want to process from the database
+            sql = """
+                SELECT room_id, count(*) FROM current_state_events
+                GROUP BY room_id
+            """
+            txn.execute(sql)
+            rooms = [{"room_id": x[0], "events": x[1]} for x in txn.fetchall()]
+
+            self._simple_insert_many_txn(txn, TEMP_TABLE, rooms)
+
+            progress["stage"] = "processing"
+            self._background_update_progress_txn(
+                txn, "populate_user_directory", progress
+            )
+
+        def _delete_staging_area(txn):
+            sql = "DROP TABLE IF EXISTS " + TEMP_TABLE
+            txn.execute(sql)
+
+        if progress.get("stage") != "processing":
```

Comment:
Instead of having stages I'd probably have them has three separate background updates, i.e. "fill out temp table", "populate user dir" and "delete temp table".

Reply:
good idea, done